### DISCUSSION TITLE
Can I use xpm install offline？

### DISCUSSION THREAD

**Discussion by andy27182** [xpack/xpm-js#182] [Q&A] [answered]
### Q: 
When I use xpm install, the download fails due to network reasons, but the zip archive can be downloaded normally through the browser, so I consider decompressing the corresponding package file in the central repository, and then use xpm install to create a link in the project . Still unsuccessful, I'm not sure if there is something wrong with my unzip path.

### ERR:
```
helloworld-arm-qemu> xpm install
@my-scope/helloworld-arm-qemu...
+ @micro-os-plus/startup@5.2.0
+ @micro-os-plus/build-helper@2.11.0
+ @micro-os-plus/diag-trace@4.1.1
+ @micro-os-plus/architecture-aarch32@3.1.0
+ @micro-os-plus/semihosting@8.0.0
+ @micro-os-plus/devices-qemu-aarch32@3.0.1
+ @xpack-dev-tools/cmake@3.21.6-1.1
+ @xpack-dev-tools/ninja-build@1.11.0-1.1
+ @xpack-dev-tools/qemu-arm@7.0.0-1.1
+ @xpack-dev-tools/arm-none-eabi-gcc@11.2.1-1.2.2
'xpacks\micro-os-plus-startup' -> 'C:\Users\xx\AppData\Roaming\xPacks\@micro-os-plus\startup\5.2.0'
'xpacks\micro-os-plus-architecture-aarch32' -> 'C:\Users\xx\AppData\Roaming\xPacks\@micro-os-plus\architecture-aarch32\3.1.0'
'xpacks\micro-os-plus-build-helper' -> 'C:\Users\xx\AppData\Roaming\xPacks\@micro-os-plus\build-helper\2.11.0'
'xpacks\micro-os-plus-diag-trace' -> 'C:\Users\xx\AppData\Roaming\xPacks\@micro-os-plus\diag-trace\4.1.1'
'xpacks\micro-os-plus-semihosting' -> 'C:\Users\xx\AppData\Roaming\xPacks\@micro-os-plus\semihosting\8.0.0'
'xpacks\micro-os-plus-devices-qemu-aarch32' -> 'C:\Users\xx\AppData\Roaming\xPacks\@micro-os-plus\devices-qemu-aarch32\3.0.1'
@xpack-dev-tools/cmake@3.21.6-1.1 => 'C:\Users\xx\AppData\Roaming\xPacks\@xpack-dev-tools\cmake\3.21.6-1.1'
@xpack-dev-tools/ninja-build@1.11.0-1.1 => 'C:\Users\xx\AppData\Roaming\xPacks\@xpack-dev-tools\ninja-build\1.11.0-1.1'
@xpack-dev-tools/arm-none-eabi-gcc@11.2.1-1.2.2 => 'C:\Users\xx\AppData\Roaming\xPacks\@xpack-dev-tools\arm-none-eabi-gcc\11.2.1-1.2.2'
Downloading https://github.com/xpack-dev-tools/cmake-xpack/releases/download/v3.21.6-1/xpack-cmake-3.21.6-1-win32-x64.zip...
Downloading https://github.com/xpack-dev-tools/ninja-build-xpack/releases/download/v1.11.0-1/xpack-ninja-build-1.11.0-1-win32-x64.zip...
@xpack-dev-tools/qemu-arm@7.0.0-1.1 => 'C:\Users\xx\AppData\Roaming\xPacks\@xpack-dev-tools\qemu-arm\7.0.0-1.1'
Downloading https://github.com/xpack-dev-tools/qemu-arm-xpack/releases/download/v7.0.0-1/xpack-qemu-arm-7.0.0-1-win32-x64.zip...
Downloading https://github.com/xpack-dev-tools/arm-none-eabi-gcc-xpack/releases/download/v11.2.1-1.2/xpack-arm-none-eabi-gcc-11.2.1-1.2-win32-x64.zip...
request to https://github.com/xpack-dev-tools/ninja-build-xpack/releases/download/v1.11.0-1/xpack-ninja-build-1.11.0-1-win32-x64.zip failed, reason: read ECONNRESET in fetch https://github.com/xpack-dev-tools/ninja-build-xpack/releases/download/v1.11.0-1/xpack-ninja-build-1.11.0-1-win32-x64.zip
If you have an aggressive antivirus, try to reconfigure it, or temporarily disable it.
error: download failed, quit
```
### OS：
- windows11

### unzip path：
```
C:\Users\xx\AppData\Roaming\xpacks:
├─@micro-os-plus
└─@xpack-dev-tools
    ├─arm-none-eabi-gcc
    │  ├─11.2.1-1.2.2
    │  │  ├─arm-none-eabi
    │  │  ├─bin
    │  │  ├─distro-info
    │  │  ├─include
    │  │  ├─lib
    │  │  ├─libexec
    │  │  └─share
    │  └─11.2.1-1.2.2.tmp
    ├─cmake
    │  ├─3.21.6-1.1
    │  │  ├─bin
    │  │  ├─distro-info
    │  │  ├─doc
    │  │  └─share
    │  └─3.21.6-1.1.tmp
    ├─ninja-build
    │  ├─1.11.0-1.1
    │  │  ├─bin
    │  │  └─distro-info
    │  └─1.11.0-1.1.tmp
    └─qemu-arm
        ├─7.0.0-1.1
        │  ├─bin
        │  ├─devices
        │  ├─distro-info
        │  ├─graphics
        │  ├─keymaps
        │  └─share
        └─7.0.0-1.1.tmp
```

**Comment by ilg-ul** [comment, marked as answer]
When xpm installs a binary package, it first downloads the npm package, then from the `package.json` file it gets the URL for the archive, and downloads it in a `.content` folder.

You can manually download the `package.json` file, manually create the `.content` folder, and move the content there.

But I recommend you try to fix your network configuration to allow npm to do this automatically. Do you use a proxy?

**Reply by andy27182** [reply]
### R1:
Thanks for your reply. I use clash for windows as the network proxy, and set up http_proxy and https_proxy, there is no problem with `curl www.google.com`, but there is still an error when executing `xpm install`, and the cause of the error becomes `HttpsProxyAgent is not a constructor in fetch ...`.

### ERR:
```
helloworld-arm-qemu> xpm install
@my-scope/helloworld-arm-qemu...
+ @micro-os-plus/architecture-aarch32@3.1.0
+ @micro-os-plus/semihosting@8.0.0
+ @micro-os-plus/devices-qemu-aarch32@3.0.1
+ @micro-os-plus/diag-trace@4.1.1
+ @micro-os-plus/startup@5.2.0
+ @micro-os-plus/build-helper@2.11.0
+ @xpack-dev-tools/cmake@3.21.6-1.1
+ @xpack-dev-tools/arm-none-eabi-gcc@11.2.1-1.2.2
+ @xpack-dev-tools/qemu-arm@7.0.0-1.1
+ @xpack-dev-tools/ninja-build@1.11.0-1.1
'xpacks\micro-os-plus-architecture-aarch32' -> 'C:\Users\xx\AppData\Roaming\xPacks\@micro-os-plus\architecture-aarch32\3.1.0'
'xpacks\micro-os-plus-startup' -> 'C:\Users\xx\AppData\Roaming\xPacks\@micro-os-plus\startup\5.2.0'
'xpacks\micro-os-plus-devices-qemu-aarch32' -> 'C:\Users\xx\AppData\Roaming\xPacks\@micro-os-plus\devices-qemu-aarch32\3.0.1'
'xpacks\micro-os-plus-semihosting' -> 'C:\Users\xx\AppData\Roaming\xPacks\@micro-os-plus\semihosting\8.0.0'
'xpacks\micro-os-plus-diag-trace' -> 'C:\Users\xx\AppData\Roaming\xPacks\@micro-os-plus\diag-trace\4.1.1'
'xpacks\micro-os-plus-build-helper' -> 'C:\Users\xx\AppData\Roaming\xPacks\@micro-os-plus\build-helper\2.11.0'
@xpack-dev-tools/arm-none-eabi-gcc@11.2.1-1.2.2 => 'C:\Users\xx\AppData\Roaming\xPacks\@xpack-dev-tools\arm-none-eabi-gcc\11.2.1-1.2.2'
@xpack-dev-tools/cmake@3.21.6-1.1 => 'C:\Users\xx\AppData\Roaming\xPacks\@xpack-dev-tools\cmake\3.21.6-1.1'
@xpack-dev-tools/qemu-arm@7.0.0-1.1 => 'C:\Users\xx\AppData\Roaming\xPacks\@xpack-dev-tools\qemu-arm\7.0.0-1.1'
Downloading https://github.com/xpack-dev-tools/cmake-xpack/releases/download/v3.21.6-1/xpack-cmake-3.21.6-1-win32-x64.zip...
HttpsProxyAgent is not a constructor in fetch https://github.com/xpack-dev-tools/cmake-xpack/releases/download/v3.21.6-1/xpack-cmake-3.21.6-1-win32-x64.zip
If you have an aggressive antivirus, try to reconfigure it, or temporarily disable it.
error: download failed, quit
```

### R2::
So I'm still trying the offline installation method, I checked the `AppData\Roaming\xpacks\@xpack-dev-tools\qemu-arm\ `directory (as an example) , There is a `package.json` file in the `\AppData\Roaming\xpacks\@xpack-dev-tools\qemu-arm\7.0.0-1.1.tmp` directory, and my node environment has always worked well, indicating that npm can be used normally. Could you please explain which level of directory the `.content` folder should be placed in. Or it is best to give an example of the location of the binary package installed under Windows OS.

**Reply by ilg-ul** [reply]
> there is no problem with curl www.google.com ...

Things are a bit more complex. Please see: https://github.com/xpack/xpm-js/issues/80

Is curl able to download from a https url?

> the offline installation method
> There is a package.json file in the \AppData\Roaming\xpacks\@xpack-dev-tools\qemu-arm\7.0.0-1.1.tmp directory,

If the 7.0.0-1.1.tmp folder looks like a npm package, and you manually downloaded the archive in the  7.0.0-1.1 folder, rename this second as .content, move it inside 7.0.0-1.1.tmp, and rename it to 7.0.0-1.1.

In the end it should look like this:

```
ilg@wksi qemu-riscv % tree -a -L 2 7.0.0-1.1
7.0.0-1.1
├── .content
│   ├── README.md
│   ├── bin
│   ├── distro-info
│   ├── include
│   ├── libexec
│   └── share
├── CHANGELOG.md
├── LICENSE
├── README.md
└── package.json

6 directories, 5 files
```

**Reply by andy27182** [reply]
> If the 7.0.0-1.1.tmp folder looks like a npm package, and you manually downloaded the archive in the 7.0.0-1.1 folder, rename this second as .content, move it inside 7.0.0-1.1.tmp, and rename it to 7.0.0-1.1.

Thanks, now `xpm install` can work.

> Things are a bit more complex. Please see: https://github.com/xpack/xpm-js/issues/80

My problem is not exactly the same as others, I checked the configuration of npm, my npm uses a mirror, no proxy is configured. therefore before using xpm, the environment variables http_proxy and https_proxy were configured in the terminal, but according to my proxy log, it seems that there is no new connection in the process of xpm. I still have to try this problem.

**Reply by ilg-ul** [reply]
> now xpm install can work

I doubt that without a proper proxy configuration you'll get very far.

**Reply by andy27182** [reply]
Yes, I will continue to work on this problem.
